src/bluesearch/database/mesh.py

```diff
@@ -0,0 +1,241 @@
+    """
+
+    def __init__(self, tree_number_to_label: dict[str, str]) -> None:
+        self.tree_number_to_label = tree_number_to_label
+        self.label_to_tree_numbers: dict[str, list[str]] = {}
+        for tree_number, label in tree_number_to_label.items():
+            if label not in self.label_to_tree_numbers:
+                self.label_to_tree_numbers[label] = []
+            self.label_to_tree_numbers[label].append(tree_number)
+
+    @classmethod
+    def load(cls, path: pathlib.Path | str) -> MeSHTree:
+        """Initialise the MeSH tree from a JSON file.
+
+        Parameters
+        ----------
+        path
+            The path to the JSON file containing the MeSH tree data. See
+            the `tree_number_to_label` parameter of the `MeSHTree`
+            constructor for the data specification.
+
+        Returns
+        -------
+        MeSHTree
+            An initialised instance of the MeSHTree.
+        """
+        with open(path) as fh:
+            tree_number_to_label = json.load(fh)
+        return cls(tree_number_to_label)
+
+    @staticmethod
+    def parents(tree_number: str) -> Generator[str, None, None]:
+        """Generate all parent tree numbers.
+
+        For example, given the tree number `H01.158.273` the parent tree
+        numbers are `H01.158` and `H01`.
+
+        Parameters
+        ----------
+        tree_number
+            A MeSH tree number, e.g. `H01.158.273`.
+
+        Yields
+        ------
+        The tree numbers of all parents of the given tree number.
+        """
+        parts = tree_number.split(".")
+        for n in reversed(range(1, len(parts))):
+            yield ".".join(parts[:n])
+
+    def parent_topics(self, topic: str) -> set[str]:
+        """Find all parent topic labels of a given topic.
+
+        Note that a topic label does not have to be unique and can be
+        assigned to multiple tree numbers. This method resolves all
+        parent topics from all tree numbers that have the given label.
+
+        Parameters
+        ----------
+        topic
+            A MeSH topic label.
+
+        Returns
+        -------
+        list
+            All parent topic labels.
+        """
+        parent_topics = set()
+        for tree_number in self.label_to_tree_numbers[topic]:
+            for parent in self.parents(tree_number):
+                parent_topics.add(self.tree_number_to_label[parent])
+
+        return parent_topics
+
+
+def resolve_parents(topics: Iterable[str], mesh_tree: MeSHTree) -> set[str]:
+    """Enhance the topic list by parents of all given topics.
+
+    Parameters
+    ----------
+    topics
+        A collection of MeSH topics.
+    mesh_tree
+        An instance of `MeSHTree`.
+
+    Returns
+    -------
+    set[str]
+        A set with the input topics and all their parent topics.
+    """
+    resolved = set(topics)
+    for topic in topics:
+        resolved |= mesh_tree.parent_topics(topic)
+
+    return resolved
```

Comment:
Yes, that's a good one. It's maybe a bit subtle, but in the current implementation it's the responsibility of `TopicInfo` to make sure its `TopicInfo.json` method is deterministic. And I think it is because of
```python
    @staticmethod
    def _add_topics(
        mapping: dict[str, list[str]], kind: str, new_topics: Iterable[str]
    ) -> None:
        """Add topics to a mapping with collection of topics.

        Parameters
        ----------
        mapping
            A mapping of the form kind -> list-of-topics that shall be
            updated in-place. For example ``{"MeSH": ["topic 1", "topic 2"]}``.
        kind
            The topic kind. Corresponds to a key in ``mapping``.
        new_topics
            The topics to add. Corresponds to a value in ``mapping``.
        """
        topics = mapping.get(kind, [])
        topics.extend(new_topics)
        mapping[kind] = sorted(set(topics))
```

Reply:
So I think as long as  one uses `Topic.Info.add_{article,journal}_topics` to update the topics the results should always be the same.